compute/cker/include/cker/operation/EinsumHelper/BCast.h

```diff
@@ -0,0 +1,236 @@
+// the right broadcasts to them, compute the broadcasted operation,
+// and possibly the gradients. In a nutshell, the caller is expected
+// to compute the broadcasted operation as following:
+//
+//   BCast b(x.shape(), y.shape());
+//   output = x.reshape(b.x_reshape()).broadcast(b.x_bcast())
+//            _op_
+//            y.reshape(b.y_reshape()).broadcast(b.y_bcast())
+//
+// For the gradient computation,
+//   grad_x = sum(grad * backprop_x(x, y), grad_x_reduce_idx)
+//            .reshape(x.shape())
+//   grad_y = sum(grad * backprop_y(x, y), grad_y_reduce_idx)
+//            .reshape(y.shape())
+// backprop_x and backprop_y are functionals of the binary function "op",
+// e.g.,
+//   for +, backprop_x(x, y) = backprop_y(x, y) = 1;
+//   for *, backprop_x(x, y) =  y, backprop_y(x, y) = x;
+//   for /, backprop_x(x, y) = 1/y, backprop_y(x, y) = -x/y^2;
+//
+// The multiplication in the grad * backprop_x itself is also
+// broadcasting following the same rule.
+//
+// TODO(zhifengc): Adds support for n-ary (n >= 2).
+class BCast
+{
+public:
+  // Constructs all helper shapes, following the aforementioned rules.
+  //
+  // If "fewer_dims_optimization" is set to true (the default), the
+  // implementation tries to reduce intermediate dimensions needed to be more
+  // efficient.  This is transparent to the caller.
+  //
+  // If false, all intermediate shapes (except for grad_{x,y}_reduce_idx()) have
+  // the same number of dimensions as the larger of the two inputs.
+  BCast(const std::vector<int32_t> &sx, const std::vector<int32_t> &sy,
+        const bool fewer_dims_optimization = true)
+  {
+    if (sx == sy && fewer_dims_optimization)
+    {
+      // Fast path for common case of identical shapes for sx and sy
+      int32_t elements = 1;
+      const int n = sx.size();
+      _output.resize(n);
+      for (int i = 0; i < n; i++)
+      {
+        const int32_t dim = sx[i];
+        elements *= dim;
+        _output[i] = dim;
+      }
+      result_.push_back(elements);
+      _x_reshape.push_back(elements);
+      _y_reshape.push_back(elements);
+      // grad_x_reduce_ and grad_y_reduce_ are left as empty
+    }
+    else
+    {
+      // Reverse the shape of x and y for convenience.
+      // After the reverse, 0-th is the inner-most dimension.
+      std::vector<int32_t> x = sx;
+      std::vector<int32_t> y = sy;
+      Reverse(&x);
+      Reverse(&y);
+
+      // 1-extend and align x and y so that they are the same size.
+      if (x.size() > y.size())
+      {
+        y.resize(x.size(), 1);
+      }
+      else
+      {
+        x.resize(y.size(), 1);
+      }
+
+      // Going through each dimension starting from the inner-most
+      // dimension, compares dimension of x and y. They are compatible if
+      // they are equal or either is 1.
+      enum State
+      {
+        UNKNOWN,
+        SAME,
+        X_ONE,
+        Y_ONE,
+      };
+      State prev = UNKNOWN;
```

Comment:
@hseok-oh I think `State` does not exist in tensorflow, and the following lines are modified version perhaps assuming that BCastLists⧼N⧽, where N = 2. Right?

Reply:
I get this code from tensorflow v2.1.